_developer-content/import-api/guides/sequencing-data-for-the-import-api.md

```diff
@@ -0,0 +1,204 @@
+      - title: "Sequencing and Replication Keys"
+        anchor: "sequence-replication-keys"
+        content: |
+          Think of `sequence` like a [Replication Key]({{ link.replication.rep-keys | prepend: site.baseurl }}).
+
+          For the Import API, it functions similarly to an `updated_at` property: Only records that have a `sequence` value greater than **[TODO- Confirm that this has to be greater than the saved sequence value for the TABLE and not the record. Pretty sure this is the case, but it should be confirmed.]**
+
+          For example: The record below has a `sequence` value of `1574800199000`:
+
+          ```json
+          {
+             "action":"upsert",
+             "sequence":1574800199000,
+             "data":{
+                "id":1,
+                "name":"Finn",
+                "age":15,
+                "has_magic":false
+             }
+          }
+          ```
+
+          In this example, `sequence` is a Unix epoch in milliseconds that translates to `Tuesday, November 26, 2019 8:29:59 PM GMT-05:00`.
+
+      - title: "Record sequencing, Primary Keys, and loading"
+        anchor: "record-sequencing-primary-keys"
+        content: |
+          TODO - Content about Primary Keys and de-duping.
+
+  - title: "Defining record sequences"
+    anchor: "define-record-sequences"
+    summary: "How to define a record's sequence"
+    content: |
+      {% for subsection in section.subsections %}
+      - [{{ subsection.title }}](#{{ subsection.anchor }})
+      {% endfor %}
+
+    subsections:
+      - title: "Generating sequence values"
+        anchor: "generate-sequence-values"
+        content: |
+          When developing your Import API script, you'll need to define how values for the `sequence` property are generated. 
+
+          A simple solution is just to use the current timestamp, but before doing so, consider the following:
+
+          - **How frequently are the records being updated?** Records that are updated every few milliseconds can result in failure if records with identical key values are pushed simultaneously. This means that records with the same Primary Key values cannot be sent during the same clock resolution.
+
+             For example: If the resolution is measured in milliseconds, records with identical Primary Key values cannot be sent during the same millisecond.
+
+          - **Are records being pushed from multiple sources?** If so, the time clocks of these sources must be synchronized. This is especially important if different sources are pushing records to the same table.
+
+      - title: "Tracking a table's maximum sequence"
+        anchor: "track-table-maximum-sequence"
+        content: |
+          In addition to generating sequence values for records, you'll also need to track the maximum sequence value for each table being pushed to the Import API. Sequence values need to continually increase or records in the destination won't be updated with new data.
```

Comment:
I'm not sure that it's really true that you need to track the max sequence value for the table. In fact I doubt most of our own integrations do that. I believe most of our integrations just use either the timestamp in millis or some function of the timestamp in millis. Here's where target-stitch adds the sequence number. It basically uses the timestamp at which it received the batch, times 1,000, as a starting point, and then auto-increments after that. So for example if the current time is 123 milliseconds, it would number the records in the batch 123000, 123001, 123002, etc...

The important thing is that the sequence number for a single primary key increases over time.

Reply:
This is a weird one - I've seen support conversations about this that were fairly recent. Any ideas who we should run this by?

Reply:
Yup, this is the line in `target-stitch` that describes Mike's message. There's no need to track with this method, since it's automatically offset using the time + index of the record inthe the current batch (zero-padded) https://github.com/singer-io/target-stitch/blob/master/target_stitch/__init__.py#L421

Reply:
@erinkcochran87 I'm pretty sure what I said originally in this thread is true. Although this stuff is tricky and it's actually implemented separately in each loader (though as far as I know they generally strive to have the same behavior). If you send me a link to a support conversation I can try to reconcile that with my understanding of how sequence numbers work.

Reply:
:+1: I can't find the support conversation at the moment, but if I happen to stumble across it at some point, I'll send it your way. I think for now we can take this out, since it may be too technical for general uses anyway.